### DISCUSSION TITLE
Allow var keyword for later assigned values

### DISCUSSION THREAD

**Discussion by eduarddejong** [dotnet/csharplang#6200] [Language Ideas]
Hi everyone! I am very new as someone who writes anything on github, but I got a very simple idea for here, and I also didn't find this same idea by anyone else yet:

Would it be a possibility for the compiler to simply infer a variable it's type based on later assignments?
The reason I am asking this is because I like things to be more consistent in code, when using the `var` keyword on most places.

Currently we can only use the `var` keyword for any directly assigned local variable, but not when a declared variable needs to get it's value later, like in conditional situations.
Then we suddenly need to declare it's type explicitly (making code inconsistent and ugly), or to make an extra initial assignment with an unused value (unnecessary and inefficient).

Let's start off with an example how this improvement could look:
```
var text;
if (args[0] == "date")
{
    text = $"{DateTime.Now:yyyy-MM-ss}";
}
else if (args[0] == "time")
{
    text = $"{DateTime.Now:HH:mm:ss}";
}
else
{
    throw new Exception("Need to specify first arg");
}
Console.WriteLine($"Result: {text}");
```

And I know that a ternary operator (?:) could have been used in this situation, but there are more than enough situations where this is not the case, for example when a lot of conditional lines of code are executed instead of just a single assignment.

Well, ok - but how?:
As long as the assigned type in all situations is the same, or at least compatible (like inherited types), I think a compiler should be able to figure it out.

**Comment by HaloFour** [comment]
See: #520

Clarity issues aside it's not hard to see where having even a few infers variables with initializers that refer to each other, possibly multiple times, could result in code that it both unreadable and very expensive for the compiler to interpret.

**Comment by DavidArno** [comment]
> And I know that a ternary operator (?:) could have been used in this situation, but there are more than enough situations where this is not the case, for example when a lot of conditional lines of code are executed instead of just a single assignment.

With modern pattern matching, so many "multiple test" use cases can be covered with a single expression now, eg your above code could be expressed as:
```c#
var text = args switch {
    ["date",..] => $"{DateTime.Now:yyyy-MM-ss}",
    ["time", ..] => $"{DateTime.Now:HH:mm:ss}",
    _ => null
};

Console.WriteLine($"Result: {text}");
```
which opens up the possibility of testing some really quite complex combinations of arguments. Further, "a lot of conditional lines of code are executed instead of just a single assignment" ought to be setting off alarm bells as that makes for complex code. If for example, if each of those "lots of ... lines of code" were refactored into a separate method, you could again use pattern matching to select the correct method and then invoke it, decoupling the selection expressions from the executing code and so simplifying things:
```c#
var method = args switch {
    ["date",..] => DateMethod,
    ["time", ..] => TimeMethod
    _ => DefaultMethod
};

var text = method();

string DateMethod()
{
    //lots of lines of code
    return $"{DateTime.Now:yyyy-MM-ss}";
}

etc
```

**Comment by eduarddejong** [comment]
Thanks for replying! The point about the switch expressions and pattern matching is clear and I would almost agree, but I consider the following...

One important thing I forgot to mention is the situation of conditionally assigning not just 1, but multiple variables at the same time.
And with this respect the switch expression is just not so much different from ternary operator situation.
Both support methods (or anonymous/lambda functions) to be used inside them, but this works all that simple as long as it's about returning just 1 value.

Otherwise we still fall back to if and switch statements in the old style, or we will really need repeated use of those ternary operators / modern style switch espressions.

Consider the following example:
```
if (args.Length < 3)
{
    return;
}
var date;
var time;
try
{
    switch (args[0])
    {
        case "local":
            date = DateTime.ParseExact(args[1], "d", CultureInfo.CurrentCulture);
            time = DateTime.ParseExact(args[2], "t", CultureInfo.CurrentCulture);
            break;
        case "invariant":
            date = DateTime.ParseExact(args[1], "d", CultureInfo.InvariantCulture);
            time = DateTime.ParseExact(args[2], "t", CultureInfo.InvariantCulture);
            break;
        default:
            return;
    }
}
catch (FormatException)
{
    Console.Error.WriteLine("Nope, that didn't work. (:");
    return;
}
Console.WriteLine($"Date: {date:yyyy-MM-ss}");
Console.WriteLine($"Time: {time:HH:mm:ss}");
```

Of course if we really want we could make extra objects or tuples to combine multiple values inside 1 thing, but that looks still a workaround to me, not a true solution.
Additionally to that, we have a try-catch block here. And as far as I know there's still no expression version of that today (would be great as well though).

**Reply by theunrepentantgeek** [reply]
> we have a try-catch block here. And as far as I know there's still no expression version of that today

It's come up a number of times - and all the proposals to date have ~founded~ foundered on the twin rocks of ambiguity and predictability.

* Implicit try scope when catch or finally block is given #908
* Allow singleline expressions in try-catch blocks #4676
* Allow try without braces #6192

Essentially, the `{` braces `}` serve to make the language unambiguous; eliding them would require introducing a bunch of special case rules that everyone would have to learn and remember - rules around which there is little to no agreement.

**Reply by jnm2** [reply]
Foundered?

**Reply by theunrepentantgeek** [reply]
Thanks! #blush #edited

**Reply by eduarddejong** [reply]
Thanks, I got it. Meanwile I have also been searching around about this subject. And I think I should have left the sentence out of my comment, it's totally another topic with it's own complexity and problems.
But I would like to add that I agree that changing things about the try-catch is very close to nonsense as it tends to cause more problems than it solves. While solving problems is the general idea behind coding. :)
And I also realize in many cases it's just a way of looking for a solution when already doing things wrong, like trying to solve to many things with try-catch statements instead of just simply relying on throwings (when you better catch them in higher scopes).

The thought with which I added this last sentence was still focussed on the point of not always being able to use expression syntax, and therefore having to declare variables without directly initializing them, making the _var_ syntax impossible with the today's language rules.

**Comment by eduarddejong** [comment]
Maybe I need to add something extra here:

1. I am not sure myself about this idea, that's why I decided to start this discussion.

2. Though the motivation for it was really the experience looking at my own code and seeing `var` keywords on most places these days, but missing them in the special cases of uninitialized variables.
When all variables start with var, my idea was that I can always start focussing on their name. And this requires you to use clear naming of course - as always should be done. A variable name should say something about it's type.

3. I have actually never felt fully sure about the `var` keyword usage in general. During the first years I literally refused to use it, since my opinion was that variables their type should always be explicitly mentioned, especially for readability reasons as well.
The motivation to change my opinion later was driven by arguments about removing the redundancy in initializations (e.g. `List<string> myList = new List<string>()`), having variable names themselves to be informative about their type, the benefits of intellisense that tells you everything and finally seeing more and more people in the world using the `var` keyword as well.

So current mindset tends to go for either _always_ using the keyword where possible, or literally the opposite in _never_ using it.

**Comment by eduarddejong** [comment]
See my new reply in #520
